File: src/1_two_sum/solution.py
# https://leetcode.com/problems/two-sum/

from typing import List

## TC: O(N)
## SC: O(N)
def method(nums: List[int], target: int) -> List[int]:
    hash_map = {}
    for i, num in enumerate(nums):
        diff = target - num
        if diff not in hash_map:
            hash_map[num] = i
        else:
            return [i, hash_map[diff]]
    return []


if __name__ == "__main__":
    print(method([2, 7, 11, 15], 9))  ## [1, 0]

File: src/1_Two_Sum/solution.ts
// https://leetcode.com/problems/two-sum/

// TC: O(N)
// SC: O(N)
function twoSum(nums: number[], target: number): number[] {
  const hash_map = new Map();
  for (let i = 0; i <= nums.length; i++) {
    const diff = target - nums[i];
    if (hash_map.has(diff)) {
      return [i, hash_map.get(diff)];
    }
    hash_map.set(nums[i], i);
  }
  return [];
}

console.log(twoSum([2, 7, 11, 15], 9)); // [1, 0]
